src/protocol/boolean/bit_decomposition.rs

```diff
@@ -38,49 +40,170 @@ impl BitDecomposition {
+        // Step 2.1: Edge case, if r is coincidentally a bit decomposition of [a]_p, we're done.
+        if c == F::ZERO {
+            return Ok(r.b_b);
+        }
 
-        // Step 6. q = d >=? p (note: the paper uses p <? d, which is incorrect)
-        let q_p = BitwiseLessThanPrime::greater_than_or_equal_to_prime(
-            ctx.narrow(&Step::IsPLessThanD),
+        // Step 3. Compute [q]_p = 1 - [r <_B p - c ]_p. q is 1 iff r + c >= p in the integers.
+        let p_minus_c: u128 = F::PRIME.into() - c.as_u128();
+        let q_p = bitwise_less_than_constant(
+            ctx.narrow(&Step::IsRLessThanPMinusC),
             record_id,
-            &d_b,
+            &r.b_b,
+            p_minus_c.into(),
         )
         .await?;
 
-        // Step 7. a bitwise scalar value `f_B = bits(2^el - p)`
+        // Step 4 has a lot going on. We'd going to break it down into substeps.
+        // Step 4.1. Make a bitwise scalar value of f = 2^el + c - p.
+        // let el = usize::try_from(u128::BITS - F::PRIME.into().leading_zeros()).unwrap();
         let el = u128::BITS - F::PRIME.into().leading_zeros();
-        let x = (1 << el) - F::PRIME.into();
+        let two_exp_el = u128::pow(2, el);
+        let _f_int: u128 = two_exp_el + c.as_u128() - F::PRIME.into();
+        debug_assert!(_f_int < u64::max_value().into());
+        let f_int: u64 = _f_int as u64;
+
+        // Step 4.2. Compute [g]_B = (f_i - c_i) [q]_p + c_i
+        let mut g_bin: Vec<G> = Vec::with_capacity(usize::try_from(el).unwrap());
+        for bit_index in 0..el {
+            // these are single bits, so let's make them bools
+            let f_i: bool = (f_int >> bit_index) & 1 == 1;
+            let c_i: bool = (c.as_u128() >> bit_index) & 1 == 1;
+            // g_i can either be c_i (known to be either 0, 1), or either [q]_p, ¬[q]_p
+            // using an enum here because we're going to do something cleaver in bitwise add
+            let g_i: G = match (f_i, c_i) {
+                // Case where f_i - c_i == 0
+                (false, false) | (true, true) => match c_i {
+                    true => G::One,
+                    false => G::Zero,
+                },
+                // Case where f_i - c_i = 1
+                (true, false) => G::Q,
+                // Case where f_i - c_i = -1
+                (false, true) => G::NotQ,
+                // this is what g_i should actually be here, so we can use this down the road
+                // let g_i_foo = S::share_known_value(&ctx, F::ONE) - &q_p.clone();
+            };
+            g_bin.push(g_i);
+            // let f_i_minus_c_i: u64 = f_i - c_i;
+            // let g_prime: S = q_p.clone().into() * f_i_minus_c_i;
+            // // c_i ∈ {0, 1}, so F::truncate_from will work
+            // debug_assert!(c_i <= 1_u128);
+            // g_bin.push(S::share_known_value(&ctx, F::truncate_from(c_i)) + &g_prime);
+        }
+
+        // Step 5. Compute BitwiseSum([r]_B, [g]_B)
```

Comment:
that deserves to be a sub protocol?

Reply:
it probably does. I've refactored it into a function, but I'm not sure if it should actually go into a new file. the `GBIterator` we create is very contextual to this specific protocol for bit decomposition, and not a generic bitwise addition, so it probably makes sense to keep it "close".